File: frontend/src/utils/textWrapper.ts
// frontend/src/utils/textWrapper.ts
/**
 * Smart Text-Wrapping Utility for SAP Master Data Migration (TypeScript / React).
 * Enforces word boundaries and character length constraints for Name and Street fields.
 */

export const NAME_LIMITS: Record<string, number> = {
  Name1: 40,
  Name2: 40,
};

export const STREET_LIMITS: Record<string, number> = {
  Street1: 60,
  Street2: 40,
  Street3: 40,
  Street4: 40,
  Street5: 40,
};

export function splitTextByWordBoundary(
  text: string | null | undefined,
  limits: number[] | Record<string, number>
): string[] | Record<string, string> {
  const isDictMode = typeof limits === 'object' && !Array.isArray(limits);
  let fieldKeys: string[] = [];
  let fieldLimits: number[] = [];

  if (isDictMode) {
    fieldKeys = Object.keys(limits as Record<string, number>);
    fieldLimits = fieldKeys.map((k) => (limits as Record<string, number>)[k]);
  } else {
    fieldLimits = [...(limits as number[])];
  }

  if (fieldLimits.length === 0) {
    return isDictMode ? {} : [];
  }

  const rawText = (text || '').trim();
  if (!rawText) {
    const emptyArray = fieldLimits.map(() => '');
    if (isDictMode) {
      const res: Record<string, string> = {};
      fieldKeys.forEach((k) => (res[k] = ''));
      return res;
    }
    return emptyArray;
  }

  const words = rawText.split(/\s+/);
  const results: string[] = [];

  let currentFieldIdx = 0;
  let currentWords: string[] = [];
  let currentLen = 0;
  const totalCapacity = fieldLimits.reduce((a, b) => a + b, 0);

  for (const word of words) {
    if (currentFieldIdx >= fieldLimits.length) {
      console.warn(
        `Input text exceeds total combined capacity (${totalCapacity} chars). Overflow text beginning with '${word}' was truncated.`
      );
      break;
    }

    let currentFieldLimit = fieldLimits[currentFieldIdx];
    const addedLen = currentLen === 0 ? word.length : word.length + 1;

    if (currentLen + addedLen <= currentFieldLimit) {
      currentWords.push(word);
      currentLen += addedLen;
    } else {
      if (currentWords.length > 0) {
        results.push(currentWords.join(' '));
        currentWords = [];
        currentLen = 0;
        currentFieldIdx++;

        if (currentFieldIdx >= fieldLimits.length) {
          console.warn(
            `Input text exceeds total combined capacity (${totalCapacity} chars). Overflow text beginning with '${word}' was truncated.`
          );
          break;
        }
      }

      currentFieldLimit = fieldLimits[currentFieldIdx];
      if (word.length <= currentFieldLimit) {
        currentWords.push(word);
        currentLen = word.length;
      } else {
        // Single word exceeds field limit
        console.warn(
          `Single word '${word.substring(0, 15)}...' (${word.length} chars) exceeds field limit (${currentFieldLimit} chars). Force splitting word.`
        );
        let remWord = word;
        while (remWord && currentFieldIdx < fieldLimits.length) {
          const cLimit = fieldLimits[currentFieldIdx];
          const part = remWord.substring(0, cLimit);
          remWord = remWord.substring(cLimit);
          if (remWord) {
            results.push(part);
            currentFieldIdx++;
            currentWords = [];
            currentLen = 0;
          } else {
            currentWords = [part];
            currentLen = part.length;
          }
        }
        if (remWord) {
          console.warn(
            `Input text exceeds total combined capacity (${totalCapacity} chars). Overflow text was truncated.`
          );
          break;
        }
      }
    }
  }

  if (currentWords.length > 0 && currentFieldIdx < fieldLimits.length) {
    results.push(currentWords.join(' '));
  }

  while (results.length < fieldLimits.length) {
    results.push('');
  }

  const finalResults = results.map((res, i) => res.trim().substring(0, fieldLimits[i]));

  if (isDictMode) {
    const resDict: Record<string, string> = {};
    fieldKeys.forEach((key, i) => {
      resDict[key] = finalResults[i];
    });
    return resDict;
  }

  return finalResults;
}

export function splitName(nameText: string | null | undefined): Record<string, string> {
  return splitTextByWordBoundary(nameText, { Name1: 40, Name2: 40 }) as Record<string, string>;
}

export function splitStreet(streetText: string | null | undefined): Record<string, string> {
  return splitTextByWordBoundary(streetText, {
    Street1: 60,
    Street2: 40,
    Street3: 40,
    Street4: 40,
    Street5: 40,
  }) as Record<string, string>;
}

/**
 * Automatically applies smart text wrapping on record objects (for Name and Street fields).
 */
export function applySmartTextWrappingToRecord(record: Record<string, string>): Record<string, string> {
  const updated = { ...record };

  // 1. Identify Name field candidates
  const nameKeys = ['NAME', 'NAME1', 'NAME_FIRST', 'NAME_ORG1', 'Name', 'Name 1', 'Name1', 'Customer Name', 'Supplier Name', 'Vendor Name'];
  const name2Keys = ['NAME2', 'NAME_LAST', 'NAME_ORG2', 'Name 2', 'Name2'];

  const foundNameKey = nameKeys.find((k) => updated[k] && updated[k].trim().length > 0);
  if (foundNameKey) {
    const rawName = updated[foundNameKey];
    const nameDict = splitName(rawName);
    updated[foundNameKey] = nameDict.Name1;

    const foundName2Key = name2Keys.find((k) => k in updated) || 'NAME2';
    if (nameDict.Name2) {
      updated[foundName2Key] = nameDict.Name2;
    }
  }

  // 2. Identify Street / Address field candidates
  const street1Keys = ['STREET', 'STREET1', 'STR_SUPPL1', 'Street', 'Street 1', 'Street1', 'Address', 'Address 1', 'Address1', 'Full Address'];
  const street2Keys = ['STREET2', 'STR_SUPPL2', 'Street 2', 'Street2', 'Address 2', 'Address2'];
  const street3Keys = ['STREET3', 'STR_SUPPL3', 'Street 3', 'Street3', 'Address 3', 'Address3'];
  const street4Keys = ['STREET4', 'Street 4', 'Street4', 'Address 4', 'Address4'];
  const street5Keys = ['STREET5', 'Street 5', 'Street5', 'Address 5', 'Address5'];

  const foundStreetKey = street1Keys.find((k) => updated[k] && updated[k].trim().length > 0);
  if (foundStreetKey) {
    const rawStreet = updated[foundStreetKey];
    const streetDict = splitStreet(rawStreet);
    updated[foundStreetKey] = streetDict.Street1;

    const key2 = street2Keys.find((k) => k in updated) || 'STREET2';
    const key3 = street3Keys.find((k) => k in updated) || 'STREET3';
    const key4 = street4Keys.find((k) => k in updated) || 'STREET4';
    const key5 = street5Keys.find((k) => k in updated) || 'STREET5';

    if (streetDict.Street2 && updated[key2] !== undefined) updated[key2] = streetDict.Street2;
    if (streetDict.Street3 && updated[key3] !== undefined) updated[key3] = streetDict.Street3;
    if (streetDict.Street4 && updated[key4] !== undefined) updated[key4] = streetDict.Street4;
    if (streetDict.Street5 && updated[key5] !== undefined) updated[key5] = streetDict.Street5;
  }

  return updated;
}

File: core/text_wrapper.py
# core/text_wrapper.py
"""
Smart Text-Wrapping Utility for SAP Master Data Migration.

Handles word-boundary enforcement and length constraints when mapping long 
strings (e.g., Vendor/Customer Names and Address Streets) into SAP master data fields.
"""

import logging
import pandas as pd
from typing import List, Dict, Union, Optional

logger = logging.getLogger(__name__)

# Exact SAP Standard Character Limit Configurations
NAME_LIMITS: Dict[str, int] = {
    "Name1": 40,
    "Name2": 40,
}

STREET_LIMITS: Dict[str, int] = {
    "Street1": 60,
    "Street2": 40,
    "Street3": 40,
    "Street4": 40,
    "Street5": 40,
}


def split_text_by_word_boundary(
    text: Optional[str],
    limits: Union[List[int], Dict[str, int]]
) -> Union[List[str], Dict[str, str]]:
    """
    Splits an input string across multiple fields based on maximum character limits,
    enforcing word boundaries so that words are not split in half (unless an individual 
    word exceeds a field's maximum character limit).

    Args:
        text (str | None): Raw input string to split.
        limits (List[int] | Dict[str, int]): Character limits for each field,
            e.g. [40, 40] or {"Name1": 40, "Name2": 40}.

    Returns:
        List[str] | Dict[str, str]: Slices of text mapped to field limits. 
            Returns dict if `limits` is a dict, or list if `limits` is a list.
    """
    is_dict_mode = isinstance(limits, dict)
    if is_dict_mode:
        field_keys = list(limits.keys())
        field_limits = [limits[k] for k in field_keys]
    else:
        field_keys = None
        field_limits = list(limits)

    if not field_limits:
        return {} if is_dict_mode else []

    if text is None:
        text = ""

    text = text.strip()
    if not text:
        empty_list = [""] * len(field_limits)
        if is_dict_mode:
            return {key: "" for key in field_keys}
        return empty_list

    words = text.split()
    results: List[str] = []
    
    current_field_idx = 0
    current_words: List[str] = []
    current_len = 0
    total_capacity = sum(field_limits)

    for word in words:
        if current_field_idx >= len(field_limits):
            logger.warning(
                f"Input text exceeds total combined capacity ({total_capacity} chars). "
                f"Overflow text beginning with '{word}' was truncated."
            )
            break

        current_field_limit = field_limits[current_field_idx]
        added_len = len(word) if current_len == 0 else len(word) + 1

        if current_len + added_len <= current_field_limit:
            current_words.append(word)
            current_len += added_len
        else:
            # Word does not fit in the current field.
            if current_words:
                results.append(" ".join(current_words))
                current_words = []
                current_len = 0
                current_field_idx += 1

                if current_field_idx >= len(field_limits):
                    logger.warning(
                        f"Input text exceeds total combined capacity ({total_capacity} chars). "
                        f"Overflow text beginning with '{word}' was truncated."
                    )
                    break

            # Try to place the word in the new/current field
            current_field_limit = field_limits[current_field_idx]
            if len(word) <= current_field_limit:
                current_words.append(word)
                current_len = len(word)
            else:
                # Edge case: Single word length exceeds the max limit of the current field.
                logger.warning(
                    f"Single word '{word[:15]}...' ({len(word)} chars) exceeds field limit "
                    f"({current_field_limit} chars). Force splitting word across field boundaries."
                )
                rem_word = word
                while rem_word and current_field_idx < len(field_limits):
                    c_limit = field_limits[current_field_idx]
                    part = rem_word[:c_limit]
                    rem_word = rem_word[c_limit:]
                    if rem_word:
                        results.append(part)
                        current_field_idx += 1
                        current_words = []
                        current_len = 0
                    else:
                        current_words = [part]
                        current_len = len(part)

                if rem_word:
                    logger.warning(
                        f"Input text exceeds total combined capacity ({total_capacity} chars). "
                        f"Overflow text was truncated."
                    )
                    break

    if current_words and current_field_idx < len(field_limits):
        results.append(" ".join(current_words))

    # Pad missing fields with empty strings
    while len(results) < len(field_limits):
        results.append("")

    # Strip and enforce maximum bounds as safeguard
    final_results = [res.strip()[:field_limits[i]] for i, res in enumerate(results)]

    if is_dict_mode:
        return {field_keys[i]: final_results[i] for i in range(len(field_keys))}
    return final_results


def split_name(name_text: Optional[str]) -> Dict[str, str]:
    """
    Convenience wrapper for Vendor/Customer Name fields.
    Name1: 40 chars
    Name2: 40 chars
    """
    return split_text_by_word_boundary(name_text, NAME_LIMITS)  # type: ignore


def split_street(street_text: Optional[str]) -> Dict[str, str]:
    """
    Convenience wrapper for Vendor/Customer Address/Street fields.
    Street1: 60 chars
    Street2: 40 chars
    Street3: 40 chars
    Street4: 40 chars
    Street5: 40 chars
    """
    return split_text_by_word_boundary(street_text, STREET_LIMITS)  # type: ignore


def apply_text_wrapping_to_dataframe(
    df: pd.DataFrame,
    source_col: str,
    target_limits: Dict[str, int]
) -> pd.DataFrame:
    """
    Integrates text wrapping into a Pandas DataFrame.
    Splits `source_col` across target columns specified in `target_limits`.

    Example:
        df = apply_text_wrapping_to_dataframe(df, 'Full_Address', STREET_LIMITS)
    """
    if source_col not in df.columns:
        logger.warning(f"Source column '{source_col}' not found in DataFrame.")
        for field in target_limits:
            df[field] = ""
        return df

    # Apply split function row-by-row
    split_dicts = df[source_col].apply(lambda text: split_text_by_word_boundary(text, target_limits))
    
    # Expand dictionary into separate columns
    split_df = pd.DataFrame(list(split_dicts), index=df.index)
    
    for col in target_limits.keys():
        df[col] = split_df[col]

    return df
